17224 - EFT Payment reminder notifications

ochiu opened the pull request:
*Issue #:*
https://github.com/bcgov/entity/issues/17224

*Description of changes:*
- added unpaid statement reminder job for 7 day reminder and payment due reminder
- Included process to update invoices to overdue status

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

seeker25 commented on jobs/payment-jobs/tasks/unpaid_statement_notify_task.py:
```diff
@@ -0,0 +1,146 @@
+    def _notify_for_monthly(cls):  # pylint: disable=too-many-locals
+        """Notify for unpaid monthly statements with an amount owing."""
+        # Get the latest monthly statement
+        now = datetime.now()
```
Note: datetime.now() might be different locally versus the UTC version on the container

ochiu replied:
Updated to use existing current_local_time() helper which uses timezone config

seeker25 commented on jobs/payment-jobs/tasks/unpaid_statement_notify_task.py:
```diff
@@ -0,0 +1,146 @@
+    @classmethod
+    def get_statement_owing(cls, auth_account_id: str, statement_id: str):
+        """Get statement owing amount by account id."""
+        result = db.session.query(func.sum(InvoiceModel.total - InvoiceModel.paid).label('total_due')) \
```
This job might be specific to for EFT - check with Ethan

ochiu replied:
Confirmed with Ethan that this is currently specific to EFT.

Are you suggesting the query to be updated to specifically target the EFT invoices?

seeker25 replied:
Yes

ochiu commented on jobs/payment-jobs/tasks/unpaid_statement_notify_task.py:
```diff
@@ -0,0 +1,146 @@
+    @classmethod
+    def _get_account(cls, auth_token: str, auth_account_id: str):
+        """Return account contact by calling auth API."""
+        get_account_endpoint = current_app.config.get('AUTH_API_ENDPOINT') + f'orgs/{auth_account_id}'
```
As discussed, this call doesn't seem to be necessary as a deactivated org(account) results in a deleted payment account

seeker25 commented on jobs/payment-jobs/tasks/unpaid_statement_notify_task.py:
```diff
@@ -0,0 +1,131 @@
+from utils.mailer import publish_payment_notification
+
+
+class UnpaidStatementNotifyTask:
```
Can we add in a blurb, noting this is for EFT only?

ochiu replied:
Added.

seeker25 commented on jobs/payment-jobs/tasks/common/enums.py:
```diff
@@ -30,3 +30,9 @@ class PaymentDetailsGlStatus(Enum):
     CMPLT = 'CMPLT'
+
+
+class OrgStatus(Enum):
```
Not sure if we need this anymore

ochiu replied:
Good catch, no longer needed and removed

seeker25 commented on jobs/payment-jobs/tasks/unpaid_statement_notify_task.py:
```diff
@@ -0,0 +1,131 @@
+        now = current_local_time()
+        previous_month = now.replace(day=1) - timedelta(days=1)
+
+        send_notification = False
```
candidate for a function from 52 -> 64

seeker25 commented on jobs/payment-jobs/tasks/unpaid_statement_notify_task.py:
```diff
@@ -0,0 +1,146 @@
+        return query.first()
+
+    @classmethod
+    def get_statement_owing(cls, auth_account_id: str, statement_id: str):
```
As part of this job, should we set the invoices to OVERDUE?

ochiu commented on jobs/payment-jobs/tasks/unpaid_statement_notify_task.py:
```diff
@@ -0,0 +1,131 @@
+            .join(PaymentAccountModel) \
+            .join(StatementInvoicesModel) \
+            .filter(PaymentAccountModel.auth_account_id == auth_account_id) \
+            .filter(InvoiceModel.invoice_status_code.in_((InvoiceStatus.SETTLEMENT_SCHEDULED.value,
```
Since we are sending out reminders 7 days before and on the due date, I don't believe they are technically OVERDUE yet, so I included the additional statuses based on the unpaid invoice notify job.

ochiu commented on jobs/payment-jobs/tasks/unpaid_statement_notify_task.py:
```diff
@@ -0,0 +1,131 @@
+                    to_emails = ','.join([str(recipient.email) for recipient in recipients])
+
+                    # Send payment notification if payment account is using EFT and there is an amount owing
+                    if payment_account.payment_method == PaymentMethod.EFT.value and summary['total_due'] > 0:
```
moved payment method and summary total check up before pulling recipients

ochiu commented:
Latest updates from dicusssions:
- updated requirements pay-api dependency to support latest changes and resolve job CI failing
- Included process to update invoices to overdue status
- renamed job as it doesn't just deal with notifications alone - also checks for overdue
- removed custom query for checking total owing (now uses the pay-api model with updates to support per statement owing)
- added / cleaned up tests
